features/steps/validation_handling.py

```diff
@@ -0,0 +1,281 @@
+
+def check_layer_for_entity_instance(i, stack_tree):
+    for layer in stack_tree:
+        if len(layer) > i and layer[i] and isinstance(layer[i], ifcopenshell.entity_instance):
+            return layer[i]
+    return None
+
+def get_activation_instances(context):
+    """Returns the activation instances of the current context. To be used for 'attribute stacking', e.g. in GEM004"""
+    stack_tree = get_stack_tree(context)
+    return [check_layer_for_entity_instance(i, stack_tree) for i in range(len(stack_tree[0]))]
+
+def validate_step(step_text):
+    def wrapped_step(func):
+        return step(step_text)(execute_step(func))
+
+    return wrapped_step
+
+def execute_step(fn):
+    @wraps(fn)
+    #@todo gh break function down into smaller functions
+    def inner(context, **kwargs):
+        step_type = context.step.step_type
+        if step_type.lower() == 'given': # behave prefers lowercase, but accepts both
+            try:
+                next(fn(context, **kwargs), None)
+            except TypeError:
+                pass
+        elif step_type.lower() == 'then':
+            if not getattr(context, 'applicable', True):
+                validation_outcome = ValidationOutcome(
+                    outcome_code=ValidationOutcomeCode.N00010,  # "NOT_APPLICABLE", Given statement with schema/mvd check failed
+                    observed=None,
+                    expected=None,
+                    feature=context.feature.name,
+                    feature_version=misc.define_feature_version(context),
+                    severity=OutcomeSeverity.NA,
+                    check_execution_id=check_execution_id
+                )
+                context.gherkin_outcomes.append(validation_outcome)
+            else:
+                instances = getattr(context, 'instances', None) or (context.model.by_type(kwargs.get('entity')) if 'entity' in kwargs else [])
+
+                #exclude empty list from instances or no stack_tree; e.g. in case of SPS001
+                try:
+                    activation_instances = get_activation_instances(context) if instances and get_stack_tree(context) else instances
+                except TypeError: #e.g. with IFC001
+                    activation_instances = instances
+
+
+                validation_outcome = ValidationOutcome(
+                    outcome_code=ValidationOutcomeCode.X00040,  # "Executed", but not no error/pass/warning
+                    observed=None,
+                    expected=None,
+                    feature=context.feature.name,
+                    feature_version=misc.define_feature_version(context),
+                    severity=OutcomeSeverity.EXECUTED,
+                    check_execution_id=check_execution_id
+                )
+                context.gherkin_outcomes.append(validation_outcome)
+
+                for i, inst in enumerate(instances):
+                    activation_inst = inst if activation_instances==instances else activation_instances[i]
```

Comment:
> activation_instances is instances in some cases.

In many cases indeed. `Context.instances` are often actual ifc entity instances, but not always; we have some rules  in which we have an eleborate stack_tree build up by multiple context.instances which sometimes end with a string value. We however want to inject an actual ifc instance in the database instead of a string. 

https://github.com/buildingSMART/ifc-gherkin-rules/blob/53c803f2838ee311d1e949d4af61a796d3f1525a/features/GEM004_Constraints-on-representation-identifiers.feature#L10-L15

In this case, instances is [None, None, ('Body', 'Axis')]. We discussed, correct me if I'm wrong, we should inject the last inst in the stack_tree that is actually an ifc instance. So activation_instances in this case is` [IfcSite, IfcBuilding, IfcProductDefinitionShape] `for which the latter is injected to the db. With an outcomeSeverity of either 'pass', 'fail' or 'warning'. 

`Do we know activation_instances is always same length as instances so that we do not get a IndexError?
`
I think we do, since activation is derived from the stack_tree and every layer here has the same length. In case the list of context.instances is shorter than the previous one, we're adding a None in that layer in context.instances (as is the case in ` [None, None, ('Body', 'Axis')].`]. In case there are more values, they are encapsulated in a tuple within the layer (as is also the case in the example with 'Body' and 'Axis'). 
Just to be sure, I've checked this manually with a couple of testfiles and this seems consistent.

Reply:
Thanks, could you add this as a comment to the code though :)

Reply:
Done. Also added some docstrings to the `get_activation_instances` function. Seems better than having all the info in these conversations. [add documentation for activation instances](https://github.com/buildingSMART/ifc-gherkin-rules/pull/119/commits/a3e4b0babe7b34d691525dda717bd979bcd8c106)
I'm not completely sure about stating that the latest instances in the stack_tree is the instance that will be reported. For example 
```
    [(None), ('Value', 'Axis')]
    [(IfcShapeRepr, IfcShapeRepr), (IfcShapeRepr, IfcShapeRepr)]
    [IfcProductDefintionShape, IfcProductDefinitionShape]
    [IfcRoof, IfcSlab]
```
Wouldn't it be better to link the rule to `IfcRoof` and `IfcSlab` instances instead of `IfcProdDefShape` ?

Perhaps we could do this by looking at the previous steps (scenario + background) 
`[s.name for s in (context.feature.background.steps + context.scenario.steps) if s.keyword == 'Given']
`
```
'A model with Schema "IFC4.3"',
'An IfcAlignmentSegment',
'Its attribute Representation',
'Its attribute Representations',
'Its attribute RepresentationIdentifier',
'The value must be "Axis"',
```
Then we can get a list of the strings that are entity instances of the schema of the model (there probably is a better way that I'm not aware of)
`entity_names = [ent.name() for ent in ifcopenshell.ifcopenshell_wrapper.schema_by_name(context.model.schema).entities()]`

Then we can look for a Given statement 'An IfcEntity' and get the index. 

-  Reverse step_names because the step names are in reversed order as in the stack_tree (values of attribute RepresentationIdentifier are on index '0' in the stack_tree, whereas they are at index '4' in set of step.names'
- Look for a combination of 'An' and a string that is an ifc_instance in the schema

```
index = (next((index for index, step in enumerate(list(reversed(step_names))) 
                 if 'An' in step.split() and any(entity in step for entity in entity_names))) -1 )
```
Then we can get the relevant instances out of the stack_tree
`get_stack_tree(context)[index] # returns [IfcRoof, IfcSlab])`

Have to test this further. Note it's only for attribute stacking (so 'Given an IfcEntity - attributes'), in other rules it is,  so far, more straightforward.

Reply:
Really nice ideas, another avenue to explore would be to somehow define that in the gherkin statements, which layer to retain as the activation. 

> I'd also do this in a seperate PR

+💯

Reply:
> I'd also do this in a separate PR

@Ghesselink afraid of hitting 100? 😆

Reply:
Hehe, busted ;)